File: app/src/main/java/com/hm/viewdemo/widget/LinearGradientView.java
package com.hm.viewdemo.widget;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.LinearGradient;
import android.graphics.Paint;
import android.graphics.Shader;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;

import com.hm.viewdemo.R;

/**
 * Created by dumingwei on 2020/5/7
 * <p>
 * Desc: 绘制View渐变背景LinearGradient，从左到右水平渐变
 * 可以只传入一个颜色，则不会渐变
 */
public class LinearGradientView extends View {

    private static final String TAG = "LinearGradientView";

    private Paint paint = new Paint();
    private LinearGradient backGradient;
    private int startColor;
    private int endColor;

    private Color color;

    public LinearGradientView(Context context) {
        this(context, null);
    }

    public LinearGradientView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public LinearGradientView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs);
    }

    private void init(Context context, AttributeSet attrs) {
        TypedArray ta = context.obtainStyledAttributes(attrs, R.styleable.LinearGradientView);
        startColor = ta.getInt(R.styleable.LinearGradientView_start_color, R.color.vip_start_color);
        endColor = ta.getInt(R.styleable.LinearGradientView_end_color, R.color.vip_end_color);
        ta.recycle();
    }

    /**
     * @param color 不用渐变
     */
    public void setSameColor(int color) {
        this.startColor = color;
        this.endColor = color;
        backGradient = new LinearGradient(0, 0, getWidth(), 0,
                new int[]{startColor, endColor}, null, Shader.TileMode.CLAMP);
        invalidate();
    }

    public void setStartColor(int startColor) {
        this.startColor = startColor;
        backGradient = new LinearGradient(0, 0, getWidth(), 0,
                new int[]{startColor, endColor}, null, Shader.TileMode.CLAMP);
        invalidate();
    }


    public void setEndColor(int endColor) {
        this.endColor = endColor;
        backGradient = new LinearGradient(0, 0, getWidth(), 0,
                new int[]{startColor, endColor}, null, Shader.TileMode.CLAMP);
        invalidate();
    }

    public void setColors(int startColor, int endColor) {
        this.startColor = startColor;
        this.endColor = endColor;
        backGradient = new LinearGradient(0, 0, getWidth(), 0,
                new int[]{startColor, endColor}, null, Shader.TileMode.CLAMP);
        invalidate();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        //获取View的宽高
        int width = getWidth();
        int height = getHeight();
        paint.setShader(null);
        paint.setShader(backGradient);
        canvas.drawRect(0, 0, width, height, paint);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        Log.i(TAG, "onSizeChanged: ");
        super.onSizeChanged(w, h, oldw, oldh);
        //从左到右水平变化
        backGradient = new LinearGradient(0, 0, getWidth(), 0,
                new int[]{startColor, endColor}, null, Shader.TileMode.CLAMP);
    }
}

File: app/src/main/java/com/hm/viewdemo/custom_view/chapter7/widget/LinearGradientView.kt
package com.hm.viewdemo.custom_view.chapter7.widget

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.util.Log
import android.view.View

/**
 * Created by dumingwei on 2019-08-27.
 * Desc:
 */
class LinearGradientView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {

    companion object {
        private const val DEFAULT_SIZE = 200
    }

    private val TAG = "LinearGradientView"

    private var paint: Paint = Paint()
    private var mMatrix = Matrix()
    private var tileMode: Shader.TileMode = Shader.TileMode.CLAMP

    init {
        setLayerType(LAYER_TYPE_SOFTWARE, null)
        paint.textSize = 80f
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val widthSpecModel = View.MeasureSpec.getMode(widthMeasureSpec)
        val widthSpecSize = View.MeasureSpec.getSize(widthMeasureSpec)

        val heightSpecModel = View.MeasureSpec.getMode(heightMeasureSpec)
        val heightSpecSize = View.MeasureSpec.getSize(heightMeasureSpec)

        if (widthSpecModel == View.MeasureSpec.AT_MOST && heightSpecModel == View.MeasureSpec.AT_MOST) {
            setMeasuredDimension(DEFAULT_SIZE, DEFAULT_SIZE)
        } else if (widthSpecModel == View.MeasureSpec.AT_MOST) {
            setMeasuredDimension(DEFAULT_SIZE, heightSpecSize)
        } else if (heightSpecModel == View.MeasureSpec.AT_MOST) {
            setMeasuredDimension(widthSpecSize, DEFAULT_SIZE)
        }
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        if (w > 0 && h > 0) {
            manyColorGradient(w, h)
        }
    }


    /**
     * 多个颜色渐变
     */
    private fun manyColorGradient(w: Int, h: Int) {

        val colors = intArrayOf(Color.RED, Color.GREEN, Color.BLUE)
        val positions = floatArrayOf(0f, 0.7f, 1f)

        val linearGradient = LinearGradient(
            0f, 0f, w * 1.0f, 0f,
            colors, positions, tileMode
        )

        paint.shader = linearGradient
    }

    override fun onDraw(canvas: Canvas) {
        drawGradientText(canvas)

    }

    private var tran = 0
    private var advance = 5
    private var text = "Android绘图小糊涂,Android绘图小糊涂,Android绘图小糊涂,Android绘图小糊涂,Android绘图小糊涂,Android绘图小糊涂"

    private var rect = Rect()

    //绘制渐变文字
    private fun drawGradientText(canvas: Canvas) {
        val colors = intArrayOf(Color.BLACK, Color.RED)
        //val colors = intArrayOf(0xff000000.toInt(), 0x7f000000, 0x0f000000)
        paint.getTextBounds(text, 0, text.length, rect)

        val textWidth = rect.width()

        Log.i(TAG, "drawGradientText: $textWidth")
        val linearGradient = LinearGradient(
            0f, 0f, -width * 1.0f, 0f,
            colors, null, Shader.TileMode.CLAMP
        )

        mMatrix.setTranslate(tran.toFloat(), 0f)
        linearGradient.setLocalMatrix(mMatrix)

        tran += advance

        if (tran >= textWidth * 2) {
            tran = 0
        }

        paint.shader = linearGradient
        canvas.drawText(text, 0f, height / 2f, paint)
        invalidate()
    }

}